mistral-search-engine.ipynb

```diff
@@ -0,0 +1,721 @@
+      ]
+     },
+     "execution_count": 21,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "function_result = tools_to_str(names_to_functions[function_name](**function_params))\n",
+    "function_result"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "\"At the Apple Worldwide Developers Conference (WWDC) in 2024, Apple made several significant announcements related to artificial intelligence (AI). The tech giant unveiled a suite of new AI capabilities to be integrated into its newest operating system. A notable development was the collaboration with OpenAI's ChatGPT, which aimed to enhance the functionality of Siri, Apple's interactive voice feature. This integration would allow Siri to provide more relevant and contextual information across various apps on Apple devices.\\n\\nThe partnership with OpenAI was seen as a major move to stay competitive with other industry leaders like Google and Microsoft, who had been emphasizing AI integration in their products. Additionally, it was announced that ChatGPT would be built into Apple's phones, computers, and iPads later in the year. This marked a significant step for Apple in the AI race, as they had been utilizing AI behind the scenes for years to power features on their devices.\\n\\nDuring the WWDC 2024, Apple also showcased updates to its own apps and operating systems, as well as new tools for developers to use in their apps. The conference was held at Apple's Cupertino, California headquarters, and demonstrated the company's commitment to advancing AI technology in its products.\""
+      ]
+     },
+     "execution_count": 22,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "messages.append(ChatMessage(role=\"tool\", name=function_name, content=function_result, tool_call_id=tool_call.id))\n",
+    "\n",
+    "response = client.chat(model=model, messages=messages)\n",
+    "response.choices[0].message.content\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Chat in a chain (cleaner user experience)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "messages = []\n",
+    "\n",
+    "while True:\n",
+    "    input_ = input(\"Ask: \")\n",
+    "    messages.append(ChatMessage(role=\"user\", content=input_))\n",
+    "    response = client.chat(model=model, messages=messages, tools=tools, tool_choice=\"auto\")\n",
```

Comment:
This might not work if the model decides to not use a function

Reply:
I can force the tool choice.